src/hangar/external/plugin_manager.py

```diff
@@ -0,0 +1,100 @@
+import pkg_resources
+from typing import Callable
+
+
+class PluginManager(object):
+    """
+    Container class that holds the information about available plugins and
+    provides required method to fetch and clean up the plugin systems
+    """
+
+    valid_provides = ['load', 'save', 'show', 'board_show']
+
+    def __init__(self):
+        self._plugin_store = {}  # ex: {'pil': loaded_pil_module}
+        self._default_plugins = {}  # ex: {'jpg': {'save': 'pil'}}
+        self.plugins_loaded = False
+
+    def reset_plugins(self):
+        """
+        Reset plugin clears the existing storages and then scans
+        `hangar.external.plugins` for plugins. Once `plugin store` is populated,
+        it creates finds the default plugins to make auto-inference based on
+        file format possible
+        """
+        self._clear_plugins()
+        self._scan_plugins()
+        self._read_defaults()
+
+    def _clear_plugins(self):
+        """
+        Clear the plugin state to the default, i.e., where no plugins are loaded
+        """
+        self._plugin_store.clear()
+        self._default_plugins.clear()
+        self.plugins_loaded = False
+
+    def _scan_plugins(self):
+        """
+        Scan for entry points, find the plugins and store them in provided storage
+        containers
+        """
+        for entry_point in pkg_resources.iter_entry_points('hangar.external.plugins'):
+            PluginClass = entry_point.load()
+            self._plugin_store[entry_point.name] = PluginClass()
+        self.plugins_loaded = True
+
+    def _read_defaults(self):
+        """
+        Populate default plugin dict that maps file formats to plugins and methods. This
+        is used to infer which plugin to use at runtime based on file format
+        """
+        for fname, plugin in self._plugin_store.items():
+            generator = ((ext, method) for ext in plugin.accepts for method in plugin.provides)
+            for pair in generator:
+                if pair not in self._default_plugins:
+                    self._default_plugins[pair] = fname
+
+    def get_plugin(self, method: str, plugin: str = None, extension: str = None) -> Callable:
+        """Load installed plugin.
+
+        User either needs to specify which plugin to load or should provide
+        file format to infer which plugin to use
+
+        Parameters
+        ----------
+        method : str
+            Which method to import from the plugin. Methods implemented by the
+            extension author should be publically announced in ``valid_provides``
```

Comment:
@rlizzo There is a slight statement. `valid_provides` is the list of `methods` `PluginManager` says valid. It's not from plugin developers. Or am I missing something here?

Reply:
Ahh my mistake. So if I'm looking at this correctly now, this should read somehting like 

```
Which method to import from the plugin. Methods implemented by the
extension author should be declared as arguments passed into the 
BasePlugin superclass constructor at the time a plugin subclass is loaded
by the hangar plugin manager?
```

(probably in simpler terms?)